### DISCUSSION TITLE
Can training step return dictionary?

### DISCUSSION THREAD

**Discussion by igoldov** [Lightning-AI/pytorch-lightning#8153] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
My training_step returns {‘loss’: loss, ‘num’: len(batch)}, so I could calculate mean loss at the end of epoch. But now I get this warning, that my training_step returns None and loss doesn't display in the progress bar. How can I return dict from training_step and display loss in the progress bar at the same tim?

**Comment by tshu-w** [comment, marked as answer]
> Can training step return dictionary?

Yes, check the [doc](https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#training-step) of `LightningModule.training_step`.

> How can I return dict from training_step and display loss in the progress bar at the same time?

just return dict like what you do now and use `self.log(loss, prog_bar=True)` to display loss in prog_bar.

> so I could calculate mean loss at the end of epoch

BTW, If you just want mean loss, I think pytorch_lightning have already done for you and you don't need to modified training_step_end(not sure)
